File: backend/src/main/java/com/locationservice/controller/ProfileController.java
package com.locationservice.controller;

import com.locationservice.dto.ProfileResponse;
import com.locationservice.dto.TelemetryResponse;
import com.locationservice.dto.UserSettingsDto;
import com.locationservice.entity.User;
import com.locationservice.service.ProfileService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api")
public class ProfileController {

    private final ProfileService profileService;

    public ProfileController(ProfileService profileService) {
        this.profileService = profileService;
    }

    @GetMapping("/profile")
    public ResponseEntity<ProfileResponse> getProfile(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(profileService.getProfile(user));
    }

    @GetMapping("/device/telemetry")
    public ResponseEntity<TelemetryResponse> getDeviceTelemetry(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(profileService.getDeviceTelemetry(user));
    }

    @GetMapping("/settings")
    public ResponseEntity<UserSettingsDto> getSettings(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(profileService.getUserSettings(user));
    }

    @PutMapping("/settings")
    public ResponseEntity<UserSettingsDto> updateSettings(
            @AuthenticationPrincipal User user,
            @RequestBody UserSettingsDto dto) {
        return ResponseEntity.ok(profileService.updateUserSettings(user, dto));
    }
}

File: backend/src/controllers/profile.controller.ts
import { Request, Response } from 'express';
import { ProfileService } from '../services/profile.service.js';

export class ProfileController {
  public static async getProfile(req: Request, res: Response): Promise<void> {
    try {
      const response = await ProfileService.getProfile(req.user!);
      res.status(200).json(response);
    } catch (error: any) {
      res.status(500).json({ error: 'Internal Server Error', message: error.message });
    }
  }

  public static async getDeviceTelemetry(req: Request, res: Response): Promise<void> {
    try {
      const response = await ProfileService.getDeviceTelemetry(req.user!);
      res.status(200).json(response);
    } catch (error: any) {
      res.status(500).json({ error: 'Internal Server Error', message: error.message });
    }
  }

  public static async getSettings(req: Request, res: Response): Promise<void> {
    try {
      const response = await ProfileService.getUserSettings(req.user!);
      res.status(200).json(response);
    } catch (error: any) {
      res.status(500).json({ error: 'Internal Server Error', message: error.message });
    }
  }

  public static async updateSettings(req: Request, res: Response): Promise<void> {
    try {
      const response = await ProfileService.updateUserSettings(req.user!, req.body);
      res.status(200).json(response);
    } catch (error: any) {
      res.status(500).json({ error: 'Internal Server Error', message: error.message });
    }
  }
}
